src/LuaTransformer.ts

```diff
@@ -3853,12 +3854,26 @@ export class LuaTransformer {
             } else if (ts.isMethodDeclaration(element)) {
                 const expression = this.transformFunctionExpression(element);
                 properties.push(tstl.createTableFieldExpression(expression, name, element));
+            } else if (ts.isSpreadAssignment(element)) {
+                if (properties.length > 0) {
+                    const tableExpression = tstl.createTableExpression(properties, expression);
+                    tableExpressions.push(tableExpression);
+                }
+                properties = [];
+                tableExpressions.push(this.transformExpression(element.expression));
             } else {
                 throw TSTLErrors.UnsupportedKind("object literal element", element.kind, expression);
             }
         });
 
-        return tstl.createTableExpression(properties, expression);
+        if (tableExpressions.length === 0) {
+            return tstl.createTableExpression(properties, expression);
+        } else {
+            const tableExpression = tstl.createTableExpression(properties, expression);
+            tableExpressions.push(tableExpression);
+
+            return this.transformLuaLibFunction(LuaLibFeature.ObjectAssign, expression, ...tableExpressions);
```

Comment:
`Object.assign` should always get a table literal as a first argument to avoid mutation of first spread element value. It probably would be better to check the type of the first element of `tableExpressions` to handle literal spread cases (check out [TypeScript output](https://www.typescriptlang.org/play/#code/BQbwBAdF4G4IYBsCuBTAXGALgJ1WAvgQJQDcAUKGPMumAGaIDOKANJNFYqhjnofqQrgoEWF1oMEzAm2rcsuFMXKURYmhknT8s8Zqat2ozhoV9lZMgGMA9gDtGmMNhR0wAXjAh8K9fN6GIi5uAr5GwRZAA)).

Reply:
Really good edge cases, thanks!